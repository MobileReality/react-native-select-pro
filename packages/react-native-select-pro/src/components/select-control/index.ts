export { SelectControl } from './select-control';
