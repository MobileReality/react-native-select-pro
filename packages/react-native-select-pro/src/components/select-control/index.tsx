import React, {
    ComponentPropsWithRef,
    forwardRef,
    ReactElement,
    useEffect,
    useRef,
    useState,
} from 'react';
import {
    AccessibilityInfo,
    Animated,
    Image,
    ImageStyle,
    Pressable,
    StyleSheet,
    Text,
    TextStyle,
    View,
    ViewStyle,
    TextInputProps,
} from 'react-native';

import {
    BORDER_WIDTH,
    COLORS,
    FONT_SIZE,
    PADDING,
    SHAPE,
} from '../../constants/styles';
import { isAndroid } from '../../helpers';
import type { OptionalToRequired } from '../../helpers/types/OptionalToRequired';
import type { OptionType, Select } from '../../index';
import { Action, DispatchType, Position, State } from '../../state/types';
import type { OnPressSelectControlType, OnSetPosition } from '../../types';
import { ClearOption } from '../clear-option';
import { MultiSelect } from '../multi-select';
import { SelectInput } from '../select-input';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'selectControlStyle'
    | 'clearable'
    | 'animated'
    | 'animationDuration'
    | 'options'
    | 'disabled'
    | 'searchable'
    | 'searchPattern'
    | 'placeholderText'
    | 'placeholderTextColor'
    | 'selectControlDisabledStyle'
    | 'selectControlButtonsContainerStyle'
    | 'hideSelectControlArrow'
    | 'multiSelection'
    | 'onSelect'
    | 'selectControlArrowImageStyle'
    | 'selectControlClearOptionA11yLabel'
    | 'selectControlOpenDropdownA11yLabel'
    | 'selectControlTextStyle'
    | 'selectControlClearOptionButtonStyle'
    | 'selectControlClearOptionButtonHitSlop'
    | 'selectControlClearOptionImageStyle'
    | 'customLeftIconSource'
    | 'customLeftIconStyles'
    | 'multiSelectionOptionStyle'
>;

type SelectControlProps = OptionalToRequired<
    {
        onPressSelectControl: OnPressSelectControlType;
    } & FromSelectComponentProps &
        Pick<
            State,
            | 'isOpened'
            | 'selectedOption'
            | 'searchValue'
            | 'selectedOptionIndex'
        > & {
            dispatch: DispatchType;
        } & Pick<Position, 'aboveSelectControl'> & {
            setPosition: OnSetPosition;
        } & {inputProps?: TextInputProps}
>;

const arrowImage = require('./../../assets/icons/chevron-down.png');

export const SelectControl = forwardRef<View, SelectControlProps>(
    (
        {
            isOpened,
            animated,
            animationDuration,
            selectControlStyle,
            selectedOption,
            onPressSelectControl,
            dispatch,
            clearable,
            options,
            disabled,
            multiSelection,
            placeholderText,
            placeholderTextColor,
            searchable,
            searchPattern,
            searchValue,
            setPosition,
            selectControlArrowImageStyle,
            selectControlDisabledStyle,
            selectControlClearOptionButtonHitSlop,
            selectControlClearOptionButtonStyle,
            selectControlClearOptionImageStyle,
            selectControlClearOptionA11yLabel,
            selectControlOpenDropdownA11yLabel,
            selectControlButtonsContainerStyle,
            multiSelectionOptionStyle,
            hideSelectControlArrow,
            onSelect,
            selectControlTextStyle,
            aboveSelectControl,
            customLeftIconSource,
            customLeftIconStyles,
            selectedOptionIndex,
            inputProps,
        },
        ref,
    ) => {
        const rotateAnimation = useRef(new Animated.Value(0)).current;

        useEffect(() => {
            if (animated) {
                Animated.timing(rotateAnimation, {
                    toValue: isOpened ? 1 : 0,
                    duration: animationDuration,
                    useNativeDriver: true,
                }).start();
            }
        }, [rotateAnimation, isOpened, animated]);

        const rotate = rotateAnimation.interpolate({
            inputRange: [0, 1],
            outputRange: ['0deg', '180deg'],
        });

        const onPressRemove = (option: OptionType | null = null) => {
            if (!disabled) {
                if (multiSelection) {
                    let removedSelectedOptions: null | OptionType[] = [];
                    removedSelectedOptions = (
                        selectedOption as OptionType[]
                    ).filter(
                        (selected) =>
                            selected.value !== (option as OptionType).value,
                    );
                    if (removedSelectedOptions.length === 0) {
                        removedSelectedOptions = null;
                    }

                    const foundIndex = options.findIndex(
                        ({ value }) => value === option?.value,
                    );

                    const resolveSelectedOptionIndex = (
                        selectedOptionIndex as number[]
                    ).filter((item) => item !== foundIndex);

                    dispatch({
                        type: Action.SelectOption,
                        payload: {
                            selectedOption: removedSelectedOptions,
                            selectedOptionIndex:
                                resolveSelectedOptionIndex?.length > 0
                                    ? resolveSelectedOptionIndex
                                    : -1,
                        },
                    });
                } else {
                    dispatch({
                        type: Action.SelectOption,
                        payload: {
                            selectedOption: null,
                            selectedOptionIndex: -1,
                        },
                    });
                    if (searchable) {
                        dispatch({
                            type: Action.SetSearchValue,
                            payload: '',
                        });
                    }
                    dispatch({
                        type: Action.SetOptionsData,
                        payload: options,
                    });
                    if (onSelect) {
                        onSelect(null, -1);
                    }
                }
            }
        };

        const [isScreenReaderEnabled, setIsScreenReaderEnabled] =
            useState(false);

        useEffect(() => {
            if (!isAndroid) {
                AccessibilityInfo.isScreenReaderEnabled().then((e) => {
                    setIsScreenReaderEnabled(e);
                });
                AccessibilityInfo.addEventListener('change', (e) => {
                    setIsScreenReaderEnabled(e);
                });
            }
        }, []);

        const isShowClearOptionButton =
            clearable && selectedOption && !isScreenReaderEnabled;
        const isShowClearOptionButtonA11y =
            clearable && selectedOption && isScreenReaderEnabled && !isAndroid;

        const renderArrowImage = (): ReactElement => {
            const accessibilityLabel = 'Arrow for opening dropdown';
            const arrow: ReactElement = animated ? (
                <Animated.Image
                    source={arrowImage}
                    style={[
                        styles.arrowIcon,
                        { transform: [{ rotate }] },
                        selectControlArrowImageStyle,
                    ]}
                />
            ) : (
                <Image
                    source={arrowImage}
                    style={[
                        styles.arrowIcon,
                        isOpened
                            ? styles.arrowIconOpened
                            : styles.arrowIconClosed,
                        selectControlArrowImageStyle,
                    ]}
                />
            );
            if (multiSelection) {
                return (
                    <Pressable
                        accessibilityLabel={accessibilityLabel}
                        onPress={disabled ? undefined : onPressSelectControl}
                    >
                        {arrow}
                    </Pressable>
                );
            }
            return arrow;
        };

        const renderMultiselect = () => {
            return (
                <MultiSelect
                    disabled={disabled}
                    dispatch={dispatch}
                    isOpened={isOpened}
                    multiSelection={multiSelection}
                    placeholderText={placeholderText}
                    placeholderTextColor={placeholderTextColor}
                    searchPattern={searchPattern}
                    searchValue={searchValue}
                    searchable={searchable}
                    selectControlStyle={selectControlStyle}
                    selectControlTextStyle={selectControlTextStyle}
                    multiSelectionOptionStyle={multiSelectionOptionStyle}
                    selectedOption={selectedOption as OptionType[]}
                    setPosition={setPosition}
                    onPressRemove={onPressRemove}
                    onPressSelectControl={onPressSelectControl}
                />
            );
        };

        const renderSelection = () => {
            const selectedOptionTyped = selectedOption as OptionType; // for proper typing
            if (searchable) {
                return (
                    <SelectInput
                        disabled={disabled}
                        dispatch={dispatch}
                        isOpened={isOpened}
                        multiSelection={multiSelection}
                        placeholderText={placeholderText}
                        placeholderTextColor={placeholderTextColor}
                        searchPattern={searchPattern}
                        searchValue={searchValue}
                        selectControlTextStyle={selectControlTextStyle}
                        selectedOption={selectedOption}
                        setPosition={setPosition}
                        onPressSelectControl={onPressSelectControl}
                        inputProps={inputProps}
                    />
                );
            }
            return (
                <Text
                    numberOfLines={1}
                    style={[
                        styles.text,
                        selectControlTextStyle,
                        {
                            color: selectedOptionTyped?.label
                                ? StyleSheet.flatten(selectControlTextStyle)
                                      ?.color || COLORS.BLACK
                                : placeholderTextColor,
                        },
                    ]}
                >
                    {selectedOptionTyped?.label || placeholderText}
                </Text>
            );
        };

        const resolveAccessibilityHint = () => {
            if (!selectedOption) {
                return undefined;
            }
            if (!multiSelection) {
                const selectedOptionTyped = selectedOption as OptionType; // for proper typing
                return `Current selected item is ${selectedOptionTyped?.label}`;
            }
            return 'You have selected multiple items';
        };

        const resolveContainer = () => {
            if (multiSelection && selectedOption) {
                return { Component: View };
            }
            return { Component: Pressable };
        };

        const { Component } = resolveContainer();

        const shouldRenderClearButton =
            isShowClearOptionButton && !multiSelection;
        const shouldRenderClearButtonA11y =
            isShowClearOptionButtonA11y && !multiSelection;

        return (
            <View style={styles.rootView}>
                <Component
                    ref={ref}
                    accessibilityHint={resolveAccessibilityHint()}
                    accessibilityLabel={
                        isOpened
                            ? ''
                            : selectControlOpenDropdownA11yLabel ||
                              'Open a dropdown'
                    }
                    style={[
                        styles.container,
                        isOpened
                            ? aboveSelectControl
                                ? styles.openedAbove
                                : styles.opened
                            : {},
                        selectControlStyle,
                        disabled
                            ? [styles.disabled, selectControlDisabledStyle]
                            : {},
                    ]}
                    onPress={
                        disabled || (multiSelection && selectedOption)
                            ? undefined
                            : onPressSelectControl
                    }
                >
                    {!!customLeftIconSource && (
                        <View
                            style={[
                                styles.leftIconWrapper,
                                styles.xIconWrapper,
                            ]}
                        >
                            <Image
                                source={customLeftIconSource}
                                style={customLeftIconStyles}
                            />
                        </View>
                    )}
                    <View
                        style={[
                            styles.press,
                            multiSelection
                                ? styles.pressMultiSelection
                                : styles.pressNormal,
                        ]}
                    >
                        {multiSelection
                            ? renderMultiselect()
                            : renderSelection()}
                    </View>
                    <View
                        style={[
                            styles.iconsContainer,
                            selectControlButtonsContainerStyle,
                        ]}
                    >
                        {shouldRenderClearButton && (
                            <ClearOption
                                disabled={disabled}
                                selectControlClearOptionA11yLabel={
                                    selectControlClearOptionA11yLabel
                                }
                                selectControlClearOptionButtonHitSlop={
                                    selectControlClearOptionButtonHitSlop
                                }
                                selectControlClearOptionButtonStyle={
                                    selectControlClearOptionButtonStyle
                                }
                                selectControlClearOptionImageStyle={
                                    selectControlClearOptionImageStyle
                                }
                                onPressRemove={onPressRemove}
                            />
                        )}
                        {!hideSelectControlArrow && renderArrowImage()}
                    </View>
                </Component>
                {shouldRenderClearButtonA11y && (
                    <View style={styles.a11IconWrapper}>
                        <ClearOption
                            disabled={disabled}
                            selectControlClearOptionA11yLabel={
                                selectControlClearOptionA11yLabel
                            }
                            selectControlClearOptionButtonHitSlop={
                                selectControlClearOptionButtonHitSlop
                            }
                            selectControlClearOptionButtonStyle={
                                selectControlClearOptionButtonStyle
                            }
                            selectControlClearOptionImageStyle={
                                selectControlClearOptionImageStyle
                            }
                            onPressRemove={onPressRemove}
                        />
                    </View>
                )}
            </View>
        );
    },
);

type Styles = {
    rootView: ViewStyle;
    container: ViewStyle;
    press: ViewStyle;
    pressMultiSelection: ViewStyle;
    pressNormal: ViewStyle;
    text: TextStyle;
    opened: ViewStyle;
    openedAbove: ViewStyle;
    disabled: ViewStyle;
    iconsContainer: ViewStyle;
    arrowIcon: ImageStyle;
    arrowIconOpened: ImageStyle;
    arrowIconClosed: ImageStyle;
    xIconWrapper: ViewStyle;
    leftIconWrapper: ViewStyle;
    a11IconWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    rootView: {
        position: 'relative',
    },
    container: {
        height: 40,
        flexDirection: 'row',
        borderRadius: SHAPE,
        borderWidth: BORDER_WIDTH,
        backgroundColor: COLORS.WHITE,
    },
    press: {
        flex: 1,
        height: '100%',
        paddingHorizontal: PADDING,
        justifyContent: 'center',
    },
    pressMultiSelection: {
        paddingRight: 40,
    },
    pressNormal: {
        paddingRight: 55,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
    text: {
        fontSize: FONT_SIZE,
        textAlign: 'left',
    },
    openedAbove: {
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0,
    },
    opened: {
        borderBottomLeftRadius: 0,
        borderBottomRightRadius: 0,
    },
    iconsContainer: {
        position: 'absolute',
        right: 8,
        top: 0,
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row',
        height: '100%',
    },
    arrowIcon: {
        width: 25,
        height: 25,
        zIndex: -1,
    },
    leftIconWrapper: {
        paddingLeft: 8,
    },
    xIconWrapper: {
        height: '100%',
        justifyContent: 'center',
    },
    arrowIconOpened: {
        transform: [{ rotate: '180deg' }],
    },
    arrowIconClosed: {
        transform: [{ rotate: '0deg' }],
    },
    a11IconWrapper: {
        position: 'absolute',
        right: -20,
        borderWidth: 1,
        height: '100%',
    },
});

SelectControl.displayName = 'SelectControl';
