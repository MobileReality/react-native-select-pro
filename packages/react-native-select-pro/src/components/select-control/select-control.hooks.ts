import { useMemo } from 'react';

import { useSelectContext } from '../../context';
import { isAndroid, selectedOptionResolver } from '../../helpers';
import { useAccessibilityScreenReader } from '../../hooks';

export const useSelectControl = () => {
    const {
        isOpened,
        clearable,
        disabled,
        multiple,
        selectedOption,
        selectContainerProps,
        hideArrow,
        selectLeftIconsProps,
        selectLeftIconImageProps,
        selectRightIconsProps,
        styles,
        aboveSelectControl,
        onPressSelectControl,
    } = useSelectContext();

    const { selectedOptionLabel } = selectedOptionResolver(selectedOption);

    const isScreenReaderEnabled = useAccessibilityScreenReader();

    const accessibilityHint = useMemo(() => {
        if (!selectedOptionLabel) {
            return;
        }
        if (!multiple) {
            return `Current selected item is ${selectedOptionLabel}`;
        }
        return 'You have selected multiple items';
    }, [selectedOptionLabel, multiple]);

    const accessibilityLabel = useMemo(
        () => (isOpened ? '' : selectContainerProps?.accessibilityLabel ?? 'Open a dropdown'),
        [isOpened, selectContainerProps?.accessibilityLabel],
    );

    const clearOptionStatus = useMemo(() => {
        const result = { showClearOption: false, showClearOptionA11y: false };

        if (!multiple && clearable && selectedOption) {
            if (!isScreenReaderEnabled) {
                result.showClearOption = true;
            } else if (!isAndroid) {
                result.showClearOptionA11y = true;
            }
        }
        return result;
    }, [clearable, isScreenReaderEnabled, multiple, selectedOption]);

    const { showClearOption, showClearOptionA11y } = clearOptionStatus;

    const { select: selectStyles } = styles ?? {};
    const {
        buttons: buttonsStyles,
        leftIcon: leftIconStyles,
        container: containerStyles,
        disabled: disabledStyles,
    } = selectStyles ?? {};

    return {
        accessibilityHint,
        accessibilityLabel,
        hideArrow,
        selectLeftIconsProps,
        selectLeftIconImageProps,
        selectRightIconsProps,
        aboveSelectControl,
        onPressSelectControl,
        selectContainerProps,
        isOpened,
        disabled,
        showClearOption,
        showClearOptionA11y,
        buttonsStyles,
        leftIconStyles,
        containerStyles,
        disabledStyles,
    };
};
