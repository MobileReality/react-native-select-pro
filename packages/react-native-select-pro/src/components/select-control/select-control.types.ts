import type { ComponentPropsWithRef } from 'react';

import type { OptionalToRequired } from '../../helpers';
import type { Select } from '../../index';
import type { DispatchType, Position, State } from '../../state/types';
import type { OnPressSelectControlType, OnSetPosition } from '../../types';
import type { SelectStyles } from '../../types/styles';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'clearable'
    | 'animation'
    | 'disabled'
    | 'searchable'
    | 'searchPattern'
    | 'textInputProps'
    | 'placeholderText'
    | 'placeholderTextColor'
    | 'hideSelectControlArrow'
    | 'multiSelection'
    | 'onSelect'
    | 'onRemove'
    | 'selectControlClearOptionA11yLabel'
    | 'selectControlOpenDropdownA11yLabel'
> &
    Pick<State, 'optionsData'>;

export type SelectControlProps = OptionalToRequired<
    {
        onPressSelectControl: OnPressSelectControlType;
    } & FromSelectComponentProps &
        Pick<State, 'isOpened' | 'selectedOption' | 'searchValue' | 'selectedOptionIndex'> & {
            dispatch: DispatchType;
        } & Pick<Position, 'aboveSelectControl'> & {
            setPosition: OnSetPosition;
        } & Pick<
            SelectStyles,
            'arrowIconStyles' | 'clearOptionStyles' | 'customLeftIconStyles' | 'selectControlStyles'
        >
>;
