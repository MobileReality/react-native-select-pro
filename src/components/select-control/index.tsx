import React, { ComponentPropsWithRef, forwardRef, ReactElement, useEffect, useRef } from 'react';
import {
    Animated,
    Image,
    ImageStyle,
    Pressable,
    StyleSheet,
    Text,
    TextStyle,
    TouchableOpacity,
    View,
    ViewStyle,
} from 'react-native';

import { BORDER_WIDTH, COLORS, FONT_SIZE, PADDING, SHAPE } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers';
import type { Select } from '../../index';
import { Action, DispatchType, Position, State } from '../../state/types';
import type { OnPressSelectControlType, OnSetPosition } from '../../types';
import { SelectInput } from '../select-input';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'selectControlStyle'
    | 'clearable'
    | 'animated'
    | 'animationDuration'
    | 'options'
    | 'disabled'
    | 'searchable'
    | 'searchPattern'
    | 'placeholderText'
    | 'selectControlDisabledStyle'
    | 'selectControlButtonsContainerStyle'
    | 'hideSelectControlArrow'
    | 'onSelect'
    | 'selectControlClearOptionA11yLabel'
    | 'selectControlOpenDropdownA11yLabel'
    | 'selectControlCloseDropdownA11yLabel'
    | 'selectControlTextStyle'
    | 'selectControlClearOptionButtonStyle'
    | 'selectControlClearOptionButtonHitSlop'
    | 'selectControlClearOptionImageStyle'
    | 'customLeftIconSource'
    | 'customLeftIconStyles'
>;

type SelectControlProps = OptionalToRequired<
    {
        onPressSelectControl: OnPressSelectControlType;
    } & FromSelectComponentProps &
        Pick<State, 'isOpened' | 'selectedOption' | 'searchValue'> & {
            dispatch: DispatchType;
        } & Pick<Position, 'aboveSelectControl'> & { setPosition: OnSetPosition }
>;

const arrowImage = require('./../../assets/icons/chevron-down.png');

export const SelectControl = forwardRef<View, SelectControlProps>(
    (
        {
            isOpened,
            animated,
            animationDuration,
            selectControlStyle,
            selectedOption,
            onPressSelectControl,
            dispatch,
            clearable,
            options,
            disabled,
            placeholderText,
            searchable,
            searchPattern,
            searchValue,
            setPosition,
            selectControlDisabledStyle,
            selectControlClearOptionButtonHitSlop,
            selectControlClearOptionButtonStyle,
            selectControlClearOptionImageStyle,
            selectControlClearOptionA11yLabel,
            selectControlOpenDropdownA11yLabel,
            selectControlCloseDropdownA11yLabel,
            selectControlButtonsContainerStyle,
            hideSelectControlArrow,
            onSelect,
            selectControlTextStyle,
            aboveSelectControl,
            customLeftIconSource,
            customLeftIconStyles,
        },
        ref,
    ) => {
        const rotateAnimation = useRef(new Animated.Value(0)).current;

        useEffect(() => {
            if (animated) {
                Animated.timing(rotateAnimation, {
                    toValue: isOpened ? 1 : 0,
                    duration: animationDuration,
                    useNativeDriver: true,
                }).start();
            }
        }, [rotateAnimation, isOpened, animated]);

        const rotate = rotateAnimation.interpolate({
            inputRange: [0, 1],
            outputRange: ['0deg', '180deg'],
        });

        const onPressRemove = () => {
            if (!disabled) {
                dispatch({
                    type: Action.SelectOption,
                    payload: null,
                });
                if (searchable) {
                    dispatch({
                        type: Action.SetSearchValue,
                        payload: '',
                    });
                }
                dispatch({
                    type: Action.SetOptionsData,
                    payload: options,
                });
                if (onSelect) {
                    onSelect(null);
                }
            }
        };

        const renderArrowImage = (): ReactElement =>
            animated ? (
                <Animated.Image
                    source={arrowImage}
                    style={[styles.arrowIcon, { transform: [{ rotate }] }]}
                />
            ) : (
                <Image
                    source={arrowImage}
                    style={[
                        styles.arrowIcon,
                        isOpened ? styles.arrowIconOpened : styles.arrowIconClosed,
                    ]}
                />
            );

        const renderSelection = () => {
            if (searchable) {
                return (
                    <SelectInput
                        disabled={disabled}
                        dispatch={dispatch}
                        isOpened={isOpened}
                        onPressSelectControl={onPressSelectControl}
                        placeholderText={placeholderText}
                        searchPattern={searchPattern}
                        searchValue={searchValue}
                        setPosition={setPosition}
                    />
                );
            }
            return (
                <Text
                    numberOfLines={1}
                    style={[
                        styles.text,
                        { color: selectedOption?.label ? COLORS.BLACK : COLORS.GRAY },
                        selectControlTextStyle,
                    ]}>
                    {selectedOption?.label || placeholderText}
                </Text>
            );
        };

        return (
            <Pressable
                accessibilityLabel={
                    isOpened
                        ? selectControlCloseDropdownA11yLabel || 'Close a dropdown'
                        : selectControlOpenDropdownA11yLabel || 'Open a dropdown'
                }
                onPress={disabled ? undefined : onPressSelectControl}
                ref={ref}
                style={[
                    styles.container,
                    isOpened ? (aboveSelectControl ? styles.openedAbove : styles.opened) : {},
                    selectControlStyle,
                    disabled ? [styles.disabled, selectControlDisabledStyle] : {},
                ]}>
                {!!customLeftIconSource && (
                    <View style={[styles.leftIconWrapper, styles.xIconWrapper]}>
                        <Image source={customLeftIconSource} style={customLeftIconStyles} />
                    </View>
                )}
                <View style={styles.press}>{renderSelection()}</View>
                <View style={[styles.iconsContainer, selectControlButtonsContainerStyle]}>
                    {clearable && selectedOption && (
                        <TouchableOpacity
                            accessibilityLabel={
                                selectControlClearOptionA11yLabel || 'Clear a chosen option'
                            }
                            disabled={disabled}
                            hitSlop={
                                selectControlClearOptionButtonHitSlop
                                    ? selectControlClearOptionButtonHitSlop
                                    : { right: 3, left: 3 }
                            }
                            onPress={onPressRemove}
                            style={[styles.xIconWrapper, selectControlClearOptionButtonStyle]}>
                            <Image
                                source={require('./../../assets/icons/x.png')}
                                style={[styles.xIcon, selectControlClearOptionImageStyle]}
                            />
                        </TouchableOpacity>
                    )}
                    {!hideSelectControlArrow && renderArrowImage()}
                </View>
            </Pressable>
        );
    },
);

type Styles = {
    container: ViewStyle;
    press: ViewStyle;
    text: TextStyle;
    opened: ViewStyle;
    openedAbove: ViewStyle;
    disabled: ViewStyle;
    iconsContainer: ViewStyle;
    arrowIcon: ImageStyle;
    arrowIconOpened: ImageStyle;
    arrowIconClosed: ImageStyle;
    xIcon: ImageStyle;
    xIconWrapper: ViewStyle;
    leftIconWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    container: {
        height: 40,
        flexDirection: 'row',
        borderRadius: SHAPE,
        borderWidth: BORDER_WIDTH,
        backgroundColor: COLORS.WHITE,
    },
    press: {
        flex: 1,
        height: '100%',
        paddingHorizontal: PADDING,
        justifyContent: 'center',
        paddingRight: 55,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
    text: {
        fontSize: FONT_SIZE,
        textAlign: 'left',
    },
    openedAbove: {
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0,
    },
    opened: {
        borderBottomLeftRadius: 0,
        borderBottomRightRadius: 0,
    },
    iconsContainer: {
        position: 'absolute',
        right: 8,
        top: 0,
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row',
        height: '100%',
    },
    arrowIcon: {
        width: 25,
        height: 25,
        zIndex: -1,
    },
    leftIconWrapper: {
        paddingLeft: 8,
    },
    xIconWrapper: {
        height: '100%',
        justifyContent: 'center',
    },
    xIcon: {
        width: 20,
        height: 20,
        zIndex: 1,
    },
    arrowIconOpened: {
        transform: [{ rotate: '180deg' }],
    },
    arrowIconClosed: {
        transform: [{ rotate: '0deg' }],
    },
});
