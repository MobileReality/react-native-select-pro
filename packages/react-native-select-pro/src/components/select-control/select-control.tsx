import React, { forwardRef } from 'react';
import type { ViewStyle } from 'react-native';
import { Image, Pressable, StyleSheet, View } from 'react-native';

import { BORDER_WIDTH, COLORS, PADDING, SHAPE } from '../../constants';
import { Arrow } from '../arrow';
import { ClearOption } from '../clear-option';
import { SelectFieldType } from '../select-field-type';

import { useSelectControl } from './select-control.hooks';

export const SelectControl = forwardRef<View>((_, ref) => {
    const {
        accessibilityHint,
        accessibilityLabel,
        onPressSelectControl,
        aboveSelectControl,
        selectLeftIconImageProps,
        selectRightIconsProps,
        selectLeftIconsProps,
        hideArrow,
        selectContainerProps,
        isOpened,
        disabled,
        showClearOptionA11y,
        showClearOption,
        buttonsStyles,
        leftIconStyles,
        containerStyles,
        disabledStyles,
        multiple,
    } = useSelectControl();

    const clearOption = <ClearOption />;

    return (
        <View style={styles.rootView}>
            <Pressable
                accessibilityHint={accessibilityHint}
                disabled={disabled}
                accessibilityState={{ disabled }}
                {...selectContainerProps}
                ref={ref}
                accessibilityLabel={accessibilityLabel}
                style={[
                    styles.container,
                    multiple ? styles.multiSelect : styles.singleSelect,
                    isOpened && (aboveSelectControl ? styles.openedAbove : styles.opened),
                    containerStyles,
                    disabled && [styles.disabled, disabledStyles],
                ]}
                onPress={onPressSelectControl}
            >
                {!!selectLeftIconImageProps?.source && (
                    <View
                        {...selectLeftIconsProps}
                        style={[styles.leftIconWrapper, styles.xIconWrapper]}
                    >
                        <Image {...selectLeftIconImageProps} style={leftIconStyles} />
                    </View>
                )}
                <SelectFieldType />
                <View {...selectRightIconsProps} style={[styles.buttonsContainer, buttonsStyles]}>
                    {showClearOption && clearOption}
                    {!hideArrow && <Arrow />}
                </View>
            </Pressable>
            {showClearOptionA11y && <View style={styles.a11IconWrapper}>{clearOption}</View>}
        </View>
    );
});

type Styles = {
    buttonsContainer: ViewStyle;
    xIconWrapper: ViewStyle;
    leftIconWrapper: ViewStyle;
    a11IconWrapper: ViewStyle;
    rootView: ViewStyle;
    container: ViewStyle;
    disabled: ViewStyle;
    openedAbove: ViewStyle;
    opened: ViewStyle;
    multiSelect: ViewStyle;
    singleSelect: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    buttonsContainer: {
        position: 'absolute',
        right: 8,
        top: 0,
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'row',
        height: '100%',
    },
    leftIconWrapper: {
        paddingLeft: 8,
    },
    xIconWrapper: {
        height: '100%',
        justifyContent: 'center',
    },
    a11IconWrapper: {
        position: 'absolute',
        right: -20,
        borderWidth: 1,
        height: '100%',
    },
    rootView: {
        position: 'relative',
        width: '100%',
    },
    container: {
        width: '100%',
        height: 40,
        flexDirection: 'row',
        borderRadius: SHAPE,
        borderWidth: BORDER_WIDTH,
        backgroundColor: COLORS.WHITE,
        paddingHorizontal: PADDING,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
    openedAbove: {
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0,
    },
    opened: {
        borderBottomLeftRadius: 0,
        borderBottomRightRadius: 0,
    },
    multiSelect: {
        paddingRight: 40,
    },
    singleSelect: {
        paddingRight: 55,
    },
});

SelectControl.displayName = 'SelectControl';
