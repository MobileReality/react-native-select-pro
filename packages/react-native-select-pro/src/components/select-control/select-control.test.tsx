import React from 'react';
import { render } from '@testing-library/react-native';

import { SelectControl } from './select-control';

describe('SelectControl', () => {
    it('should generate SelectControl snapshot', () => {
        const DATA = [{ label: 'test', value: 'test' }];
        const selectControl = render(
            <SelectControl
                ref={null}
                aboveSelectControl={false}
                animation={false}
                clearable={false}
                disabled={false}
                dispatch={() => {}}
                hideSelectControlArrow={false}
                isOpened={false}
                multiSelection={false}
                optionsData={DATA}
                placeholderText={undefined}
                placeholderTextColor={undefined}
                searchPattern={undefined}
                searchValue=""
                textInputProps={{}}
                selectControlClearOptionA11yLabel={undefined}
                selectControlOpenDropdownA11yLabel={undefined}
                selectedOption={null}
                selectedOptionIndex={-1}
                clearOptionStyles={{}}
                customLeftIconStyles={{}}
                selectControlStyles={{}}
                arrowIconStyles={{}}
                setPosition={() => {}}
                onPressSelectControl={() => {}}
                onRemove={() => {}}
            />,
        );
        expect(selectControl).toMatchSnapshot();
    });
});
