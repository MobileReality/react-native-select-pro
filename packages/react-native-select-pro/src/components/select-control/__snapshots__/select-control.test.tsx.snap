// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectControl should generate SelectControl snapshot 1`] = `
<View
  style={
    {
      "position": "relative",
    }
  }
>
  <View
    accessibilityLabel="Open a dropdown"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
        },
        false,
        undefined,
        {},
      ]
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
            "paddingHorizontal": 12,
          },
          {
            "paddingRight": 55,
          },
        ]
      }
    >
      <TextInput
        accessibilityLabel="Place text"
        editable={true}
        onChangeText={[Function]}
        onPressIn={[Function]}
        style={
          [
            {
              "fontSize": 12,
            },
            undefined,
          ]
        }
        textAlign="left"
        value=""
      />
    </View>
    <View
      style={
        [
          {
            "alignItems": "center",
            "flexDirection": "row",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "right": 8,
            "top": 0,
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          {
            "testUri": "packages/react-native-select-pro/src/assets/icons/chevron-down.png",
          }
        }
        style={
          [
            {
              "height": 24,
              "width": 24,
              "zIndex": -1,
            },
            false,
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;
